### ISSUE TITLE
Slack webhooks are not working properly.

### ISSUE THREAD

**Issue by decuple-dream** [beefproject/beef#2819]
## First Steps

1. Confirm that your issue has not been posted previously by searching here: https://github.com/beefproject/beef/issues
2. Confirm that the wiki does not contain the answers you seek: https://github.com/beefproject/beef/wiki
3. Check the FAQ: https://github.com/beefproject/beef/wiki/FAQ
4. BeEF Version: 0.5.4.0
5. Ruby Version: 3.0.3
6. Browser Details : Firefox
7. Operating System: Kali linux

## Configuration

1. Have you made any changes to your BeEF configuration? Yes
2. Have you enabled or disabled any BeEF extensions? Yes

## Steps to Reproduce

This time, I cloned BeEF to GitHub and built it from scratch. Configuring Slack wasn't too difficult, so I created a WebHook, put it in the Config.yaml file, changed it to true, and was done, but I got an error.

1. First, clone the latest BeFE-XSS GitHub source(./update-beef) and modify `/extensions/notifications/config.yaml`.
```
slack:
              enable: true
              webhook_url: "https://hooks.slack.com/services/T0357AEQZP0/B052T1A..../NPeQXIs6q...."
              channel: "#beef"    # Slack channel
              username: "beef-xss" # Username can be anything
```
3. My server is in an HTTPS (cloudflare + public domain) environment with Apache, so I'll run it after I finish the additional setup.
./beef -x
```
...
[ 2:13:30][!] [GeoIP] Could not find MaxMind GeoIP database: '/usr/share/GeoIP/GeoLite2-City.mmdb'
[ 2:13:30][*] HTTP Proxy: http://127.0.0.1:6789
[ 2:13:30][*] BeEF server started (press control+c to stop)
[ 2:13:32][!] [Notifications] Invalid Slack WebHook URL
```
4. Hmm, something's wrong, read that ruby file. (extensions/notifications/channels/slack_workspaces.rb)
```ruby
if webhook_url.include?('your_webhook_url') || !webhook_url.start_with?('https://hook\.slack.com/services/')
              print_error '[Notifications] Invalid Slack WebHook URL'
              return
            end
```
5. Try requesting Slack via CURL.
```
curl -X POST --data-urlencode "payload={\"channel\": \"#beef\", \"username\": \"webhookbot\", \"text\": \"test\", \"icon_emoji\": \":ghost:\"}" https://hooks.slack.com/services/T0357AEQZP0/B053X4V2N9W/g6e7EXvP87kStX....
```
<img width="649" alt="스크린샷 2023-04-14 오후 3 22 32" src="https://user-images.githubusercontent.com/67303734/231960008-ae59a384-949d-4b76-be37-452130d4f16c.png">

6. Delete the `slack_workspaces.rb` conditional just in case and run it.
```
[ 2:26:02][*] running on network interface: 172.17.0.1
[ 2:26:02]    |   Hook URL: http://172.17.0.1:3000/hook.js
[ 2:26:02]    |_  UI URL:   http://172.17.0.1:3000/ui/panel
[ 2:26:02][*] Public:
[ 2:26:02]    |   Hook URL: https://mydomain.is:443/hook.js
[ 2:26:02]    |_  UI URL:   https://mydomain.is:443/ui/panel
[ 2:26:02][*] RESTful API key: cd19c5386b882810b2fbcebd0462a96bd2c517bd
[ 2:26:02][!] [GeoIP] Could not find MaxMind GeoIP database: '/usr/share/GeoIP/GeoLite2-City.mmdb'
[ 2:26:02][*] HTTP Proxy: http://127.0.0.1:6789
[ 2:26:02][*] BeEF server started (press control+c to stop)
[ 2:26:04][!] [Notifications] Slack notification initialization failed: SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (unable to get local issuer certificate)
```

The other BeEF project folders I had previously configured worked fine, but I didn't make a backup of them, so I rebuilt them and something is wrong. My webhooks always start with https://hooks.slack.com.

I don't know why, it was working fine before that, so I don't know what's wrong. Thanks for making such a good program.

**Comment by bcoles** [comment]
> Hmm, something's wrong, read that ruby file.

The conditional is broken.

```diff
diff --git a/extensions/notifications/channels/slack_workspace.rb b/extensions/notifications/channels/slack_workspace.rb
index a11c5612..112c6320 100644
--- a/extensions/notifications/channels/slack_workspace.rb
+++ b/extensions/notifications/channels/slack_workspace.rb
@@ -18,7 +18,7 @@ module BeEF
             channel = @config.get('beef.extension.notifications.slack.channel')
             username = @config.get('beef.extension.notifications.slack.username')
 
-            if webhook_url.include?('your_webhook_url') || !webhook_url.start_with?('https://hook\.slack.com/services/')
+            if webhook_url.include?('your_webhook_url') || !webhook_url.start_with?('https://hook.slack.com/services/')
               print_error '[Notifications] Invalid Slack WebHook URL'
               return
             end
```

> [ 2:26:04][!] [Notifications] Slack notification initialization failed: SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (unable to get local issuer certificate)

This error implies a SSL connection failure - specifically an error with certificate validation.


> My webhooks always start with https://hooks.slack.com/.

`hooks` or `hook` ?

**Comment by decuple-dream** [comment]
Hi, @bcoles! Thanks to reply! 
<img width="461" alt="스크린샷 2023-04-14 오후 4 05 32" src="https://user-images.githubusercontent.com/67303734/231969157-c9909717-0785-4342-8ad7-d74e0afb78ab.png">
Isn't this the one?

Can you tell me how to make a beef SSL connection?

**Comment by bcoles** [comment]
> Can you tell me how to make a beef SSL connection?

BeEF is already making a SSL connection. The issue is that SSL validation is failing.

This may be due to CA certificates missing from your ruby environment and/or system.

Here are a few possible solutions:

* https://stackoverflow.com/questions/4528101/ssl-connect-returned-1-errno-0-state-sslv3-read-server-certificate-b-certificat

The slack notifications extension does not do much. It largely depends on the rushover gem.

**Comment by bcoles** [comment]
> . First of all, do I need to allow https: in web_server_imitation in Config.yaml

No


> I followed the link you sent me and tried wget http://curl.haxx.se/ca/cacert.pem followed by export SSL_CERT_FILE=$PWD/cacert.pem but I get the same error.

> I also tried adding a certificate.

No

> What am I missing?

The certificate validation is failing. Ensure your Ruby environment can access your CA certificates.

**Comment by decuple-dream** [comment]
Thank you for your kind response. @bcoles 

I know you're busy, but one last question:
does your Ruby environment have an access test procedure for CA certificates?

Is there a simple command to pass the certificate through?

**Comment by bcoles** [comment]
> does your Ruby environment have an access test procedure for CA certificates?

I don't understand your question.

> Is there a simple command to pass the certificate through?

I don't understand your question.

And I don't think you understand the issue.

You know how you must access websites with self-signed certificates with `curl` using `-k` ? Or with `wget` using `--no-check-certificate` ? This is because certificate validation failed. cURL has no way of knowing if the remote certificate is trusted because it is self signed.

This is the same problem.

BeEF is trying to connect to Slack. It does not trust the connection because the certificate is not trusted. This could be for many reasons, but the most likely reason the certificate is not trusted is because your Ruby environment is not configured with the appropriate CA bundle.

If your system cannot validate certificates then it cannot validate certificates.

You could try requiring the `certified` and hope for the best. Or you could fix your Ruby environment.

The Slack notification is extremely simply. It depends entirely on the underlying slack-notifier gem. It is just a few lines:

https://github.com/beefproject/beef/blob/eb5959a9756fb9ac84467272721bd0b390896a1e/extensions/notifications/channels/slack_workspace.rb#L13-L36

I wrote this 6 years ago. It has not been maintained or updated. Very few people use it. I have no interest in setting up a test environment to see if it still works.

**Comment by bcoles** [comment]
If you want to share the webhook URL I'll see if I can at at least initialize the notifications extension on my local BeEF install.

Here's an invite for discord if you want to share in DM: https://discord.gg/Px84xDpd

**Comment by decuple-dream** [comment]
Thank you. I'm still spending time on SSL today, I just participated in Discord. I'll try a little more on my own and if that doesn't work, I'll ask for help.

**Comment by bcoles** [comment]
Regarding the WebHook validation conditional logic, I broke this when converting the regex check to a string comparison check in aa7a6f9e644ef8da9006a28379f0193df00dba99 (removed `s` character instead of `\` character). This should be fixed in #2820.

The certificate validation issue is a separate issue related to the underlying slack-notifier gem. Refer to:

https://github.com/slack-notifier/slack-notifier/issues/135
